### PR TITLE
[WIP] Add min and max as possible avg_method in plot_surf (#2790)

### LINKED ISSUE #2790: More diverse avg_method for plot_surf to handle sparse matrices
Currently, the `avg_method` parameter in `plotting.plot_surf` only takes `mean` or `median` as arguments.
Having the possibility `min`, `max` or other functions would be useful to me, in the sense that sometimes I have some sparse surface matrices, therefore using `mean` or `median` kind of kills my plots :smile: 

### Benefits to the change

It would allow for a better representation of sparse matrices on the surface.

### Pseudocode for the new behavior, if applicable

```python
import matplotlib.pyplot as plt
from nilearn import datasets, plotting
import numpy as np

fsaverage = datasets.fetch_surf_fsaverage()
n_voxels = 10242

surface = np.zeros(n_voxels)
n = 100
surface[:n] = np.ones(n)

plotting.plot_surf(
    fsaverage.pial_left,
    surface,
    avg_method="min"
)
plt.show()
```

### PR DESCRIPTION
**Pull request by alexisthual**
This PR closes #2790.

So far, this PR adds 2 possible entries for the `avg_method` argument in `plotting.plot_surf`: `min` and `max`.
We are still considering adding the possibility to pass a custom aggregation function for `avg_method`.

### CODE REVIEW

**Review by NicolasGensollen — CHANGES_REQUESTED** [review]
Thanks @alexisthual ! :+1: 
The code looks good to me. Could you add:

- tests for this (in `nilearn/plotting/tests/test_surf_plotting.py`)
- a whatsnew entry (in `doc/whatsnew.rst`)

File: nilearn/plotting/tests/test_surf_plotting.py [hunk 1]
```diff
@@ -36,6 +36,14 @@ def test_plot_surf():
     plot_surf(mesh, bg_map=bg, colorbar=True, cbar_vmin=0,
               cbar_vmax=150, cbar_tick_format="%i")
 
+    # Plot with avg_method
+    plot_surf(mesh, surf_map=rng.standard_normal(size=mesh[0].shape[0]), avg_method='mean')
```
Comment (NicolasGensollen) [inline nilearn/plotting/tests/test_surf_plotting.py:40]: I'm wondering whether we should test explicitly that the texture we obtain results from the `avg_method` provided :thinking: 
I was thinking about something along those lines:

```python
import matplotlib.pyplot as plt
import numpy as np
from nilearn.surface import load_surf_mesh
from nilearn.surface.testing_utils import generate_surf
from numpy.testing import assert_array_equal
from nilearn.plotting.surf_plotting import plot_surf

mesh = generate_surf()
rng = np.random.RandomState(42)
bg = rng.standard_normal(size=mesh[0].shape[0])
mapp = rng.standard_normal(size=mesh[0].shape[0])
mesh_ = load_surf_mesh(mesh)
coords, faces = mesh_[0], mesh_[1]

for method in ['mean', 'median', 'min', 'max']:
    display = plot_surf(mesh, surf_map=mapp, avg_method=method)
    if method == 'mean':
        agg_faces = np.mean(mapp[faces], axis=1)
    elif method == 'median':
        agg_faces = np.median(mapp[faces], axis=1)
    elif method == "min":
        agg_faces = np.min(mapp[faces], axis=1)
    elif method == "max":
        agg_faces = np.max(mapp[faces], axis=1)
    vmin = np.min(agg_faces)
    vmax = np.max(agg_faces)
    agg_faces -= vmin
    agg_faces /= (vmax - vmin)
    cmap = plt.cm.get_cmap(plt.rcParamsDefault['image.cmap'])
    assert_array_equal(cmap(agg_faces),
                       display._axstack.as_list()[0].collections[0]._facecolors)
```

WDYT @alexisthual ?

Comment (RaphaelMeudec) [inline nilearn/plotting/tests/test_surf_plotting.py:40]: Maybe you can rely on a `pytest.mark.parametrize` to lighten the test a bit? Here is an example:

```python
@pytest.mark.parametrize("plot_surf_kwargs", [
    {},
    {"bg": np.random.RandomState(42).standard_normal(size=mesh[0].shape[0])},
    {"bg": np.random.RandomState(42).standard_normal(size=mesh[0].shape[0]), "darkness": 0.5},
    {"bg": np.random.RandomState(42).standard_normal(size=mesh[0].shape[0]), "alpha": 0.5},
])
def test_plot_surf(plot_surf_kwargs):
    mesh = generate_surf()
    # Plot mesh only
    plot_surf(mesh, **plot_surf_kwargs)
```

The idea is that `plot_surf_kwargs` will take successively the value provided in the `parametrize` object

Comment (alexisthual) [reply nilearn/plotting/tests/test_surf_plotting.py]: That's a very good idea, I added your suggestion

File: nilearn/plotting/surf_plotting.py [hunk 2]
```diff
@@ -270,6 +277,17 @@ def plot_surf(surf_mesh, surf_map=None, bg_map=None,
...
             surf_map_faces = np.median(surf_map_data[faces], axis=1)
+        elif avg_method == 'min':
+            surf_map_faces = np.min(surf_map_data[faces], axis=1)
+        elif avg_method == 'max':
+            surf_map_faces = np.max(surf_map_data[faces], axis=1)
+        elif callable(avg_method):
+            surf_map_faces = np.apply_along_axis(avg_method, 1, surf_map_data[faces])
+            ## check that surf_map_faces has the same length as face_colors
+            if surf_map_faces.shape != (face_colors.shape[0],):
+                raise ValueError(f'Array computed with the custom function from avg_method does not have the correct shape: {surf_map_faces.shape[0]} != {face_colors.shape[0]}')
```
Comment (jeromedockes) [inline nilearn/plotting/surf_plotting.py:288]: ATM nilearn still supports python 3.5 so you'll have to use string.format instead of f-strings

Comment (alexisthual) [reply nilearn/plotting/surf_plotting.py]: Good catch, thanks :)

File: nilearn/plotting/surf_plotting.py [hunk 3]
```diff
@@ -270,6 +277,24 @@ def plot_surf(surf_mesh, surf_map=None, bg_map=None,
...
+        elif avg_method == 'min':
+            surf_map_faces = np.min(surf_map_data[faces], axis=1)
+        elif avg_method == 'max':
+            surf_map_faces = np.max(surf_map_data[faces], axis=1)
+        elif callable(avg_method):
+            surf_map_faces = np.apply_along_axis(avg_method, 1, surf_map_data[faces])
+
+            ## check that surf_map_faces has the same length as face_colors
+            if surf_map_faces.shape != (face_colors.shape[0],):
+                raise ValueError('Array computed with the custom function from avg_method does not have the correct shape: {} != {}'.format(surf_map_faces.shape[0], face_colors.shape[0])
```
Comment (bthirion) [inline nilearn/plotting/surf_plotting.py:289]: beware the 80 char limit per line.

**Review by bthirion — COMMENTED** [review]
LGTM overall.

File: nilearn/plotting/surf_plotting.py (hunk 3 above)
Comment (alexisthual) [reply nilearn/plotting/surf_plotting.py]: My last commit (42cc8fb) fixes this, thanks!

**Comment by bthirion** [comment]
I'm afraid there are a few more.

File: nilearn/plotting/surf_plotting.py [hunk 4]
```diff
@@ -72,9 +72,16 @@ def plot_surf(surf_mesh, surf_map=None, bg_map=None,
     colorbar : bool, optional
         If True, a colorbar of surf_map is displayed. Default=False.
 
-    avg_method : {'mean', 'median'}, optional
+    avg_method : {'mean', 'median', 'min', 'max', custom function}, optional
         How to average vertex values to derive the face value, mean results
-        in smooth, median in sharp boundaries. Default='mean'.
+        in smooth, median in sharp boundaries, min or max for sparse matrices.
+        You can also pass a custom function which will be executed though `numpy.apply_along_axis`. Here is an example of a custom function:
```
Comment (NicolasGensollen) [inline nilearn/plotting/surf_plotting.py:78]: maybe split this line

File: nilearn/plotting/tests/test_surf_plotting.py [hunk 5]
```diff
@@ -36,6 +39,35 @@ def test_plot_surf():
...
+        agg_faces -= vmin
+        agg_faces /= (vmax - vmin)
+        cmap = plt.cm.get_cmap(plt.rcParamsDefault['image.cmap'])
+        assert_array_equal(cmap(agg_faces),
+                           display._axstack.as_list()[0].collections[0]._facecolors)
+
+    ## Try custom avg_method
+    def custom_avg_function(vertices):
+        return vertices[0] * vertices[1] * vertices[2]
+    plot_surf(mesh, surf_map=rng.standard_normal(size=mesh[0].shape[0]), avg_method=custom_avg_function)
```
Comment (NicolasGensollen) [inline nilearn/plotting/tests/test_surf_plotting.py:69]: Could you also split this one?

File: nilearn/plotting/tests/test_surf_plotting.py [hunk 6]
```diff
@@ -71,6 +103,28 @@ def test_plot_surf_error():
             mesh, surf_map=rng.standard_normal(size=(mesh[0].shape[0], 2))
         )
 
+    with pytest.raises(
+        ValueError, match="Array computed with the custom function from avg_method does not have the correct shape"
```
Comment (NicolasGensollen) [inline nilearn/plotting/tests/test_surf_plotting.py:107]: I would split all these lines. Maybe something like:

```python
with pytest.raises(
    ValueError, match=("Array computed with the custom function "
                       "from avg_method does not have the correct shape")):
```

**Review by NicolasGensollen — COMMENTED** [review]
LGTM minus the lengths of a few lines (see comments below). Thanks @alexisthual !

**Review by bthirion — APPROVED** [review]
LGTM, thx !

**Review by NicolasGensollen — APPROVED** [review]
LGTM, thanks @alexisthual !

**Comment by NicolasGensollen** [comment]
Oops, it looks like there is a conflict in `whats_new` that we should resolve first!

**Comment by alexisthual** [comment]
Should be fixed now, thanks for the heads-up!

**Comment by NicolasGensollen** [comment]
Merging! :tada:

**Pull request merged**